accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -2746,26 +2746,31 @@ To match a pattern `p` against a value `v`:
         *This type test may get elided. See "Pointless type tests and legacy
         types" below.*
 
-    2.  Let `l` be the length of the list determined by calling `length` on `v`.
+    2.  Let `l` be the length of the list determined by calling `length` on `v`
+        or `0` if `length` returns a negative number. *We treat misbehaving
```

Comment:
This works but I think we can do better in terms of runtime efficiency.  Personally I'd rather leave this step as:

2. Let `l` be the length of the list determined by calling `length` on `v`.

And then insert before the next step:

3. If the list pattern has no elements, and `l <= 0`, the match succeeds.  Otherwise...

(and leave the remaining text unchanged, other than renumbering of course).

This is indistinguishable from what you specified, and more efficient because it just replaces an `== 0` check with a `<= 0` check when the list pattern is `[]`, rather than adding a `< 0` check to *every* list pattern.

(Yes, I realize that implementers are free to implement something different from what we specify if the results are indistinguishable, but I think we should give them a good starting point.)

Reply:
> 3\. If the list pattern has no elements, and `l <= 0`, the match succeeds.

What about:

```dart
class NegativeLengthList<E> implements List<E> { ... }

main() {
  if (NegativeLengthList<String>() case [...]) {
    print('?');
  }
}
```

Does that match? An even more fun one:

```dart
main() {
  if (NegativeLengthList<String>() case [...var rest]) {
    print(rest);
  }
}
```

Should that throw an exception because it passes an invalid negative argument to `sublist()`, or should it truncate the negative length to zero and then call `sublist(0, 0)`?

Reply:
Should `NegativeLengthList() case [...]` match? Yes.
There should be no length check on the `[...]` pattern at all.

A list pattern *p*'s check could start by something like:

* Let *n* be the number of non-spread patterns in *p*
* If *p* contains a spread pattern, then:
    * If *n* > 0, check that *listLength* >= *n*
* Otherwise (no spread)
    * if *n* > 0, check that *listLength* == *n*
    * otherwise check that *listLength* >= 0

A `[...]` pattern would go through that without needing to check `listLength` at all, and `[]` would check length >= 0.

Also, for a tailing spread pattern with a match pattern, `[p1, p2, ... p3]`,
the `sublist` call should not pass an `end` argument, because it goes to the end of the list.
That makes `case [...l] => l` (list copy, the weird way!) not need the list length at all.

I don't care if we specify *listLength* as `let tmp = list.length in tmp < 0 ? 0 : tmp`.
A compiler can optimize that into the behavior above, if we make sure there is no way to see that value if it's not greater than zero anyway. 
(Which is why `[...p]` should do `list.sublist(0)`, whereas `[...p, q]` requires length > 0 anyway, so `sublist(0, list.length - 1)` will be fine.)

Reply:
> > 3. If the list pattern has no elements, and `l <= 0`, the match succeeds.
> 
> What about:
> 
> ```dart
> class NegativeLengthList<E> implements List<E> { ... }
> 
> main() {
>   if (NegativeLengthList<String>() case [...]) {
>     print('?');
>   }
> }
> ```

Oops, you're right.

> Does that match? An even more fun one:
> 
> ```dart
> main() {
>   if (NegativeLengthList<String>() case [...var rest]) {
>     print(rest);
>   }
> }
> ```
> 
> Should that throw an exception because it passes an invalid negative argument to `sublist()`, or should it truncate the negative length to zero and then call `sublist(0, 0)`?

Yeah, good corner cases.  I'm glad you didn't blindly accept my suggestion 😃 

I like the approach @lrhn is suggesting.  We need to spell out all the details, but as I understand it, it basically boils down to:
- Assume that the list is well-behaved when deciding what operations to perform on it (including not calling `length` when handling a final `...`). 
- When we actually perform those operations, instead of computing `x.length`, compute `max(0, x.length)`.  *(this is equivalent to @lrhn's suggestion `let tmp = list.length in tmp < 0 ? 0 : tmp`)*.
- Optimize where appropriate (e.g. `max(0, x.length) == 0` becomes `x.length <= 0` and `max(0, x.length) == 3` becomes `x.length == 3`).

This feels very consistent with what we decided to do for ill-behaved maps.  The magic there was coming up with a definition for the `containsKey` operation that had all the consistency properties we wanted and was easily optimizable; here we are doing the same magic on `length`.  Not surprising that @lrhn had both ideas 😃

Reply:
OK, how does the latest commit look?